### DISCUSSION TITLE
Task delay start does not seem to work

### DISCUSSION THREAD

**Discussion by Victorhlab** [arkhipenko/TaskScheduler#186] [Q&A]
I am trying to enable a task again with a delay start after it received a timeout. The idea is that is a function takes too long,
to disable it and start it after some time again. 

This is what I tried, but the task just keep on running. I receive the disable event, but the delayed start part does not work. 

void task1OnDisable() {
  if (t1.timedOut()) {
    Serial.println("Task 1 has timed out." + String(millis()) + "ms. Restarting... 10S delay before restart.");
    t1.setInterval(1 * TASK_SECOND);
    t1.setIterations(TASK_FOREVER);
    t1.setTimeout(2 * TASK_SECOND);
    t1.enableDelayed(10 * TASK_SECOND);//delay 10 seconds before restarting
  }
  else {
    Serial.println("Task 1 has been disabled");
  }
}

**Comment by arkhipenko** [comment]
Hard to say without full code.
Do you also have `t1OnEnable()` by any chance? Are you setting any intervals there? OnEnable() is called immediately even for the delayed tasks. 

Enabling tasks while OnDisable() is active, equally as disabling tasks in the OnEnable() often leads to interesting unexpected results. I have been there myself many times.

**Comment by Victorhlab** [comment]
/*
    This eaxmple illustrates the use of overall Task timeout functionality:

    Task 1 - runs every 1 seconds and times out in 10 seconds
    Task 2 - runs every 5 seconds and resets the timeout every run, so runs continuosly even though the timeout is set to 10 seconds
*/

#include <Arduino.h>
#include <FirebaseESP32.h>
 #define _TASK_TIMECRITICAL      // Enable monitoring scheduling overruns
#define _TASK_SLEEP_ON_IDLE_RUN // Enable 1 ms SLEEP_IDLE powerdowns between tasks if no callback methods were invoked during the pass 
//#define _TASK_STATUS_REQUEST    // Compile with support for StatusRequest functionality - triggering tasks on status change events in addition to time only
// #define _TASK_WDT_IDS           // Compile with support for wdt control points and task ids
// #define _TASK_LTS_POINTER       // Compile with support for local task storage pointer
// #define _TASK_PRIORITY          // Support for layered scheduling priority
// #define _TASK_MICRO_RES         // Support for microsecond resolution
// #define _TASK_STD_FUNCTION      // Support for std::function (ESP8266 ONLY)
// #define _TASK_DEBUG             // Make all methods and variables public for debug purposes
#define _TASK_INLINE         // Make all methods "inline" - needed to support some multi-tab, multi-file implementations
#define _TASK_TIMEOUT

#include <TaskScheduler.h>

Scheduler ts;

void task1Callback();
void task1OnDisable();
void task2Callback();
void task2OnDisable();
bool task1OnEnable();

Task t1(1 * TASK_SECOND, TASK_FOREVER, &task1Callback, &ts, true, &task1OnEnable, &task1OnDisable);
Task t2(5 * TASK_SECOND, TASK_FOREVER, &task2Callback, &ts, false, NULL, &task2OnDisable);

void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);

  Serial.println("TaskScheduler Timeout example");
  Serial.println("=============================");

  t1.setTimeout(2 * TASK_SECOND);
  //t2.setTimeout(10 * TASK_SECOND);

  ts.enableAll();
}

void loop() {
  // put your main code here, to run repeatedly:
  ts.execute();
}


void task1Callback() {
  Serial.print("Task 1:\t");
  Serial.print(millis());
  Serial.print(": t/out=");
  Serial.print(t1.getTimeout());
  Serial.print("\tms until t/out=");
  Serial.println( t1.untilTimeout());
  delay(5000);

}
void task1OnDisable() {
  if (t1.timedOut()) {
    Serial.println("Task 1 has timed out." + String(millis()) + "ms. Restarting... 10s delay before restart.");
    t1.setInterval(1 * TASK_SECOND);
    t1.setIterations(1);
    t1.setTimeout(2 * TASK_SECOND);
    t1.enableDelayed(10 * TASK_SECOND);//delay 10 seconds before restarting
  }
  else {
    Serial.println("Task 1 has been disabled");
  }
}

bool task1OnEnable() {
  Serial.println("Task 1 has been enabled:" + String(millis()));
  t1.enableDelayed(10 * TASK_SECOND);//Duplicate as this is also stated in OnDisable -> Has no effect even removing it in OnDisable, or here
  return true;
}
void task2Callback() {
  Serial.print("Task 2:\t");
  Serial.print(millis());
  Serial.print(": t/out=");
  Serial.print(t2.getTimeout());
  Serial.print("\tms until t/out=");
  Serial.println( t2.untilTimeout());
  t2.resetTimeout();
}
void task2OnDisable() {
  if (t2.timedOut()) {
    Serial.println("Task 2 has timed out");
  }
  else {
    Serial.println("Task 2 has been disabled");
  }
}






OUTPUT:
Task 1 has been enabled:61
Task 1: 61: t/out=2000  ms until t/out=1999
Task 2: 5062: t/out=0   ms until t/out=-1
Task 1 has timed out.5062ms. Restarting... 10s delay before restart.
Task 1 has been enabled:5062
Task 2: 5073: t/out=0   ms until t/out=-1
Task 1 has timed out.7074ms. Restarting... 10s delay before restart.
Task 1 has been enabled:7074
Task 1 has timed out.9075ms. Restarting... 10s delay before restart.
Task 1 has been enabled:9075
Task 2: 10061: t/out=0  ms until t/out=-1
Task 1 has timed out.11076ms. Restarting... 10s delay before restart.
Task 1 has been enabled:11076
Task 1 has timed out.13077ms. Restarting... 10s delay before restart.
Task 1 has been enabled:13077
Task 2: 15061: t/out=0  ms until t/out=-1
Task 1 has timed out.15078ms. Restarting... 10s delay before restart.
Task 1 has been enabled:15078
Task 1 has timed out.17079ms. Restarting... 10s delay before restart.
Task 1 has been enabled:17079
Task 1 has timed out.19080ms. Restarting... 10s delay before restart.
Task 1 has been enabled:19080
Task 2: 20061: t/out=0  ms until t/out=-1
Task 1 has timed out.21081ms. Restarting... 10s delay before restart.
Task 1 has been enabled:21081
Task 1 has timed out.23082ms. Restarting... 10s delay before restart.
Task 1 has been enabled:23082
Task 2: 25061: t/out=0  ms until t/out=-1
Task 1 has timed out.25083ms. Restarting... 10s delay before restart.
Task 1 has been enabled:25083
Task 1 has timed out.27084ms. Restarting... 10s delay before restart.
Task 1 has been enabled:27084
Task 1 has timed out.29085ms. Restarting... 10s delay before restart.
Task 1 has been enabled:29085
Task 2: 30061: t/out=0  ms until t/out=-1
Task 1 has timed out.31086ms. Restarting... 10s delay before restart.
Task 1 has been enabled:31086
Task 1 has timed out.33087ms. Restarting... 10s delay before restart.
Task 1 has been enabled:33087
Task 2: 35061: t/out=0  ms until t/out=-1
Task 1 has timed out.35088ms. Restarting... 10s delay before restart.
Task 1 has been enabled:35088
Task 1 has timed out.37089ms. Restarting... 10s delay before restart.
Task 1 has been enabled:37089
Task 1 has timed out.39090ms. Restarting... 10s delay before restart.
Task 1 has been enabled:39090
Task 2: 40061: t/out=0  ms until t/out=-1
Task 1 has timed out.41091ms. Restarting... 10s delay before restart.
Task 1 has been enabled:41091
Task 1 has timed out.43092ms. Restarting... 10s delay before restart.
Task 1 has been enabled:43092
Task 2: 45061: t/out=0  ms until t/out=-1
Task 1 has timed out.45093ms. Restarting... 10s delay before restart.
Task 1 has been enabled:45093
Task 1 has timed out.47094ms. Restarting... 10s delay before restart.
Task 1 has been enabled:47094
Task 1 has timed out.49095ms. Restarting... 10s delay before restart.
Task 1 has been enabled:49095
Task 2: 50061: t/out=0  ms until t/out=-1
Task 1 has timed out.51096ms. Restarting... 10s delay before restart.
Task 1 has been enabled:51096
Task 1 has timed out.53097ms. Restarting... 10s delay before restart.
Task 1 has been enabled:53097
Task 2: 55061: t/out=0  ms until t/out=-1
Task 1 has timed out.55098ms. Restarting... 10s delay before restart.
Task 1 has been enabled:55098
Task 1 has timed out.57099ms. Restarting... 10s delay before restart.
Task 1 has been enabled:57099
Task 1 has timed out.59100ms. Restarting... 10s delay before restart.
Task 1 has been enabled:59100
Task 2: 60061: t/out=0  ms until t/out=-1
Task 1 has timed out.61101ms. Restarting... 10s delay before restart.
Task 1 has been enabled:61101
Task 1 has timed out.63102ms. Restarting... 10s delay before restart.
Task 1 has been enabled:63102
Task 2: 65061: t/out=0  ms until t/out=-1
Task 1 has timed out.65103ms. Restarting... 10s delay before restart.
Task 1 has been enabled:65103
Task 1 has timed out.67104ms. Restarting... 10s delay before restart.
Task 1 has been enabled:67104
Task 1 has timed out.69105ms. Restarting... 10s delay before restart.
Task 1 has been enabled:69105
Task 2: 70061: t/out=0  ms until t/out=-1
Task 1 has timed out.71106ms. Restarting... 10s delay before restart.
Task 1 has been enabled:71106
Task 1 has timed out.73107ms. Restarting... 10s delay before restart.
Task 1 has been enabled:73107
Task 2: 75061: t/out=0  ms until t/out=-1
Task 1 has timed out.75108ms. Restarting... 10s delay before restart.
Task 1 has been enabled:75108
Task 1 has timed out.77109ms. Restarting... 10s delay before restart.
Task 1 has been enabled:77109
Task 1 has timed out.79110ms. Restarting... 10s delay before restart.
Task 1 has been enabled:79110
Task 2: 80061: t/out=0  ms until t/out=-1
Task 1 has timed out.81111ms. Restarting... 10s delay before restart.
Task 1 has been enabled:81111
Task 1 has timed out.83112ms. Restarting... 10s delay before restart.
Task 1 has been enabled:83112
Task 2: 85061: t/out=0  ms until t/out=-1
Task 1 has timed out.85113ms. Restarting... 10s delay before restart.
Task 1 has been enabled:85113
Task 1 has timed out.87114ms. Restarting... 10s delay before restart.
Task 1 has been enabled:87114
Task 1 has timed out.89115ms. Restarting... 10s delay before restart.
Task 1 has been enabled:89115
Task 2: 90061: t/out=0  ms until t/out=-1
Task 1 has timed out.91116ms. Restarting... 10s delay before restart.
Task 1 has been enabled:91116
Task 1 has timed out.93117ms. Restarting... 10s delay before restart.
Task 1 has been enabled:93117
Task 2: 95061: t/out=0  ms until t/out=-1
Task 1 has timed out.95118ms. Restarting... 10s delay before restart.
Task 1 has been enabled:95118
Task 1 has timed out.97119ms. Restarting... 10s delay before restart.
Task 1 has been enabled:97119
Task 1 has timed out.99120ms. Restarting... 10s delay before restart.
Task 1 has been enabled:99120
Task 2: 100061: t/out=0 ms until t/out=-1
Task 1 has timed out.101121ms. Restarting... 10s delay before restart.
Task 1 has been enabled:101121
Task 1 has timed out.103122ms. Restarting... 10s delay before restart.
Task 1 has been enabled:103122
Task 2: 105061: t/out=0 ms until t/out=-1
Task 1 has timed out.105123ms. Restarting... 10s delay before restart.
Task 1 has been enabled:105123
Task 1 has timed out.107124ms. Restarting... 10s delay before restart.
Task 1 has been enabled:107124
Task 1 has timed out.109125ms. Restarting... 10s delay before restart.
Task 1 has been enabled:109125
Task 2: 110061: t/out=0 ms until t/out=-1
Task 1 has timed out.111126ms. Restarting... 10s delay before restart.
Task 1 has been enabled:111126
Task 1 has timed out.113127ms. Restarting... 10s delay before restart.
Task 1 has been enabled:113127
Task 2: 115061: t/out=0 ms until t/out=-1
Task 1 has timed out.115128ms. Restarting... 10s delay before restart.
Task 1 has been enabled:115128
Task 1 has timed out.117129ms. Restarting... 10s delay before restart.
Task 1 has been enabled:117129
Task 1 has timed out.119130ms. Restarting... 10s delay before restart.
Task 1 has been enabled:119130
Task 2: 120061: t/out=0 ms until t/out=-1
Task 1 has timed out.121131ms. Restarting... 10s delay before restart.
Task 1 has been enabled:121131
Task 1 has timed out.123132ms. Restarting... 10s delay before restart.
Task 1 has been enabled:123132
Task 2: 125061: t/out=0 ms until t/out=-1
Task 1 has timed out.125133ms. Restarting... 10s delay before restart.
Task 1 has been enabled:125133
Task 1 has timed out.127134ms. Restarting... 10s delay before restart.
Task 1 has been enabled:127134
Task 1 has timed out.129135ms. Restarting... 10s delay before restart.
Task 1 has been enabled:129135
Task 2: 130061: t/out=0 ms until t/out=-1
Task 1 has timed out.131136ms. Restarting... 10s delay before restart.
Task 1 has been enabled:131136
Task 1 has timed out.133137ms. Restarting... 10s delay before restart.
Task 1 has been enabled:133137
Task 2: 135061: t/out=0 ms until t/out=-1
Task 1 has timed out.135138ms. Restarting... 10s delay before restart.
Task 1 has been enabled:135138
Task 1 has timed out.137139ms. Restarting... 10s delay before restart.
Task 1 has been enabled:137139
Task 1 has timed out.139140ms. Restarting... 10s delay before restart.
Task 1 has been enabled:139140
Task 2: 140061: t/out=0 ms until t/out=-1
Task 1 has timed out.141141ms. Restarting... 10s delay before restart.
Task 1 has been enabled:141141
Task 1 has timed out.143142ms. Restarting... 10s delay before restart.
Task 1 has been enabled:143142
Task 2: 145061: t/out=0 ms until t/out=-1
Task 1 has timed out.145143ms. Restarting... 10s delay before restart.
Task 1 has been enabled:145143
Task 1 has timed out.147144ms. Restarting... 10s delay before restart.
Task 1 has been enabled:147144
Task 1 has timed out.149145ms. Restarting... 10s delay before restart.
Task 1 has been enabled:149145
Task 2: 150061: t/out=0 ms until t/out=-1
Task 1 has timed out.151146ms. Restarting... 10s delay before restart.
Task 1 has been enabled:151146
Task 1 has timed out.153147ms. Restarting... 10s delay before restart.
Task 1 has been enabled:153147
Task 2: 155061: t/out=0 ms until t/out=-1
Task 1 has timed out.155148ms. Restarting... 10s delay before restart.
Task 1 has been enabled:155148
Task 1 has timed out.157149ms. Restarting... 10s delay before restart.
Task 1 has been enabled:157149
Task 1 has timed out.159150ms. Restarting... 10s delay before restart.
Task 1 has been enabled:159150
Task 2: 160061: t/out=0 ms until t/out=-1
Task 1 has timed out.161151ms. Restarting... 10s delay before restart.
Task 1 has been enabled:161151
Task 1 has timed out.163152ms. Restarting... 10s delay before restart.
Task 1 has been enabled:163152
Task 2: 165061: t/out=0 ms until t/out=-1
Task 1 has timed out.165153ms. Restarting... 10s delay before restart.
Task 1 has been enabled:165153
Task 1 has timed out.167154ms. Restarting... 10s delay before restart.
Task 1 has been enabled:167154
